### PR TITLE
fix(rest-api): Add order_by query param to TI listing APIs (#41283)

### LINKED ISSUE #41283: Airflow REST GET API call for TaskInstances returns unordered items, making paged retrieval impossible
### Apache Airflow version

Other Airflow 2 version (please specify below)

### If "Other Airflow 2 version" selected, which one?

2.9.1, but nothing in the release notes thereafter indicates this issue has been fixed.

### What happened?

When using paged retrieval:

```
https://airflow/api/v1/dags/DAG/dagRuns/rc--2024-08-02_01-30/taskInstances?limit=100&offset=0
https://airflow/api/v1/dags/DAG/dagRuns/rc--2024-08-02_01-30/taskInstances?limit=100&offset=100
https://airflow/api/v1/dags/DAG/dagRuns/rc--2024-08-02_01-30/taskInstances?limit=100&offset=200
```

the task instance results are unordered, ensuring with near certainty that all subsequent paged retrievals (with offset > 0) return some tasks already returned before, and fail to return tasks that should have been in the page.  Thus, concatenating the tasks yields the correct *number* of tasks, but some are repeated and some are missing.

Full retrieval as per

```
https://airflow/api/v1/dags/DAG/dagRuns/rc--2024-08-02_01-30/taskInstances?limit=300
```

returns each task instance only once, and returns all task instances exhaustively.

### What you think should happen instead?

Task instances from API call should either be returned in some specific order, or there should be a query string parameter allowing us users to specify a designated order (such as task instance creation / appearance date).

### How to reproduce

Just issue a taskInstances GET call with a small limit, then with an offset equivalent to the small limit.  Tasks that were on the first batch are likely to appear on the second batch too.

### Operating System

Ubuntu 22.04 container

### Versions of Apache Airflow Providers

Not relevant.

### Deployment

Official Apache Airflow Helm Chart

### Deployment details

We use a very vanilla Kubernetes.

### Anything else?

Always.

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

I cannot, but I can recommend the following: all REST endpoints that list items in a paged manner must return items in a stable sort order, or at least allow the end user to specify a stable sort order criteria and document that unordered items are returned by default.

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://github.com/apache/airflow/blob/main/CODE_OF_CONDUCT.md)

### PR DESCRIPTION
**Pull request by omkar-foss**
closes #41283.

This adds db-level sorting with order_by query param to the following TI listing APIs:
1. List task instances - `/api/v1/dags/~/dagRuns/~/taskInstances`
2. List task instances (batch) - `/api/v1/dags/~/dagRuns/~/taskInstances/list`

order_by defaults to sorting by start_date (ascending) for above mentioned 2 APIs. Please note that this does NOT change the default sorting param for the List mapped task instances API.

This also adds corresponding unit tests.

---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [newsfragments](https://github.com/apache/airflow/tree/main/newsfragments).

### CODE REVIEW

File: airflow/api_connexion/endpoints/task_instance_endpoint.py [hunk 1]
```diff
@@ -297,6 +279,31 @@ def _apply_range_filter(query: Select, key: ClauseElement, value_range: tuple[T,
...
+        return (TI.start_date.asc(), TI.map_index.asc())
+    if order_by == "-start_date":
+        return (TI.start_date.desc(), TI.map_index.asc())
+    if order_by == "end_date":
+        return (TI.end_date.asc(), TI.map_index.asc())
+    if order_by == "-end_date":
+        return (TI.end_date.desc(), TI.map_index.asc())
+    if order_by == "-map_index":
+        return (TI.map_index.desc(),)
+    return ()
```
Comment (uranusjr) [inline airflow/api_connexion/endpoints/task_instance_endpoint.py]: If we receive an invalid value here, I think an error should just be raised directly inside this function. Returning nothing is too easy for the call site to not handle correctly.

File: airflow/api_connexion/endpoints/task_instance_endpoint.py [hunk 2]
```diff
@@ -378,11 +386,15 @@ def get_task_instances(
         )
         .add_columns(SlaMiss)
         .options(joinedload(TI.rendered_task_instance_fields))
-        .offset(offset)
-        .limit(limit)
     )
+
+    order_by_params = _get_order_by_params(order_by or "start_date")
```
Comment (uranusjr) [inline airflow/api_connexion/endpoints/task_instance_endpoint.py]: This default should be mentioned somewhere.

File: airflow/api_connexion/endpoints/task_instance_endpoint.py (hunk 1 above)
Comment (omkar-foss) [reply airflow/api_connexion/endpoints/task_instance_endpoint.py]: Yes makes sense, I'll raise a `ValueError` in this function on invalid value. Thanks.

File: airflow/api_connexion/endpoints/task_instance_endpoint.py (hunk 2 above)
Comment (omkar-foss) [reply airflow/api_connexion/endpoints/task_instance_endpoint.py]: Adding this to the REST API docs. Thanks.

File: airflow/api_connexion/openapi/v1.yaml [hunk 3]
```diff
@@ -5193,6 +5194,11 @@ components:
           items:
             type: string
           description: The value can be repeated to retrieve multiple matching values (OR condition).
+        order_by:
+          type: string
+          description: |
+            The name of the field to order the results by. Prefix a field name
+            with `-` to reverse the sort order.
```
Comment (jscheffl) [inline airflow/api_connexion/openapi/v1.yaml]: I think this should also mention the version it was added.
```suggestion
            The name of the field to order the results by. Prefix a field name
            with `-` to reverse the sort order.

            *New in version 3.0.0*
```

Comment (omkar-foss) [reply airflow/api_connexion/openapi/v1.yaml]: Updated docs, thanks!

File: airflow/api_connexion/endpoints/task_instance_endpoint.py (hunk 1 above)
Comment (omkar-foss) [reply airflow/api_connexion/endpoints/task_instance_endpoint.py]: PR updated.

File: airflow/api_connexion/endpoints/task_instance_endpoint.py (hunk 2 above)
Comment (omkar-foss) [reply airflow/api_connexion/endpoints/task_instance_endpoint.py]: Keeping `order_by` to default to `map_index` (same as it has been for Get Mapped Instances) for consistency. User can specify the `order_by` param as required by referring to the docs.

**Comment by omkar-foss** [comment]
Fixed all linting issues and added all your suggestions from comments above. Please review when you get a chance, thanks.

**Review by uranusjr — APPROVED** [review]

**Review by jscheffl — APPROVED** [review]

**Review by pierrejeambrun — APPROVED** [review]
Tested locally, working fine, and code is looking good.

**Pull request merged**